crates/holochain/src/core/workflow/sys_validation_workflow.rs

```diff
@@ -43,154 +42,464 @@ mod validate_op_tests;
     workspace: Arc<SysValidationWorkspace>,
-    space: Arc<Space>,
-    network: Network,
-    sys_validation_trigger: TriggerSender,
-) -> WorkflowResult<WorkComplete> {
+    current_validation_dependencies: Arc<Mutex<ValidationDependencies>>,
+) -> WorkflowResult<OutcomeSummary> {
     let db = workspace.dht_db.clone();
     let sorted_ops = validation_query::get_ops_to_sys_validate(&db).await?;
-    let start_len = sorted_ops.len();
-    tracing::debug!("Validating {} ops", start_len);
-    // TODO questionable check for saturation. The query can return up to 10,000 results and this function
-    //      will process more than NUM_CONCURRENT_OPS ops, just not in parallel. Not necessarily true that
-    //      there will be more ops to process after this workflow run.
-    let start = (start_len >= NUM_CONCURRENT_OPS).then(std::time::Instant::now);
-    let saturated = start.is_some();
-    let cascade = workspace.full_cascade(network);
-
-    validate_ops_batch(
-        sorted_ops,
-        start,
-        {
-            let space = space.clone();
-            move |so| {
-                // Create an incoming ops sender for any dependencies we find
-                // that we are meant to be holding but aren't.
-                // If we are not holding them they will be added to our incoming ops.
-                let incoming_dht_ops_sender =
-                    IncomingDhtOpSender::new(space.clone(), sys_validation_trigger.clone());
-                let workspace = workspace.clone();
-                let cascade = cascade.clone();
-                async move {
-                    let (op, op_hash) = so.into_inner();
-                    let op_type = op.get_type();
-                    let action = op.action();
-
-                    let dependency = get_dependency(op_type, &action);
-                    let dna_def = DnaDefHashed::from_content_sync((*workspace.dna_def()).clone());
-
-                    let r =
-                        validate_op(&op, &dna_def, &cascade, Some(&incoming_dht_ops_sender)).await;
-                    r.map(|o| (op_hash, o, dependency))
+
+    if sorted_ops.is_empty() {
+        tracing::trace!(
+            "Skipping sys_validation_workflow because there are no ops to be validated"
+        );
+
+        // If there's nothing to validate then we can clear the dependencies and save some memory.
+        let mut deps = current_validation_dependencies.lock();
+        deps.clear_retained_deps();
+        deps.purge_held_deps();
+
+        return Ok(OutcomeSummary::new());
+    }
+
+    let num_ops_to_validate = sorted_ops.len();
+    tracing::debug!("Sys validating {} ops", num_ops_to_validate);
+
+    let cascade = Arc::new(workspace.local_cascade());
+    let dna_def = DnaDefHashed::from_content_sync((*workspace.dna_def()).clone());
+
+    // Forget what dependencies are currently in use
+    current_validation_dependencies.lock().clear_retained_deps();
+
+    fetch_previous_records(
```

Comment:
This replaces all the cascade calls that used to be scattered through the workflow and issues them all at once for LOCAL data only